src/macaron/slsa_analyzer/package_registry/pypi_registry.py

```diff
@@ -274,25 +275,25 @@ def get_maintainer_join_date(self, username: str) -> datetime | None:
 
 
 @dataclass
-class PyPIPackageJsonAsset:
-    """The package JSON hosted on the PyPI registry.
-
-    This class matches the AssetLocator protocol.
-    """
+class PyPIPackageJsonAsset(AssetLocator):
```

Comment:
What is the reasoning behind inheriting the `AssetLocator` protocol here?

Reply:
That's used here to trigger type checking.

Reply:
Hmmm I don't think sub-classing a protocol alone is meant to trigger the type checking.
The type checking should also be triggered at call site, where an instance of `PyPIPackageJsonAsset` is accessed.
```
def run_heuristics(asset: AssetLocator):
    ...

def run_check():
    asset_list = [PyPIPackageJsonAsset(...), OtherAsset(...)]
    invalid_one = SomeThingThatIsNotAnAsset(...)
    for asset in asset_list:
        run_heuristics(asset) # No type error.
    
    run_heuristics(invalid_one) # type error.
```
I think this serves the purpose of Protocol, where any classes that behave similar to `AssetLocator` should be allowed here. This enable us to freely extend to other type of asset without having to change the interface of `run_heuristics` (i.e change the type of the `asset` parameter).

Reply:
In general I don't think we need to subclass a Protocol as described in this PEP document - https://typing.readthedocs.io/en/latest/spec/protocol.html#explicitly-declaring-implementation

> So while it’s possible to subclass a protocol explicitly, it’s not necessary to do so for the sake of type-checking.

Reply:
No, we don't need to subclass the `Protocol`, but as I said, I have added it to trigger type checking. Right now we don't store this asset like you have shown above, so we can't trigger the type checking at usage time.

Reply:
Right, I see.
If we are using `PyPIPackageJsonAsset` as the concrete typing at usage time (i.e. in the `run_heuristics` method), then I don't think we need to subclass Protocol to enable type checking. This is because the type checker already know the concrete type `PyPIPackageJsonAsset` and can provide the necessary type checking at call site. I think there is no issue with trying to make `PyPIPackageJsonAsset` follows `AssetLocator`'s interface, however, sub-classing `AssetLocator` won't help anything in our current use case and can be safely omitted here. 
In the future when we change `run_heuristic` 's interface to use `AssetLocator` (as I demonstrated above) the type checker should not raise any error because `PyPIPackageJsonAsset` has already followed `AssetLocator` interface.

I guess the reason why we didn't notice the incorrect interface for the `download` method is that at call site, we are using the concrete type of `PyPIPackageJsonAsset`. Therefore, the type check has nothing to complain about.

Reply:
>however, sub-classing AssetLocator won't help anything in our current use case and can be safely omitted here.
That's not actually correct. By sub-classing `AssetLocator`, mypy does capture the incompatible signature of the `download` method.

Reply:
Yes I agree, if we sub-class `AssetLocator`, that will lead to a constraint on the method interfaces of `PyPIPackageJsonAsset`. Mypy will ensure that `PyPIPackageJsonAsset` must exactly follows `AssetLocator` interface. 
If we don't sub-class `AssetLocator`, the code still work and mypy will only raise error if at call site we are using `download` differently than its definition in `PyPIPackageJsonAsset`. This means that our code is still type-correct. 
I guess if you want to make sure that the definition of `download` in `PyPIPackageJsonAsset` always follows that of `AssetLocator`, I am good with sub-classing `AssetLocator`. 
It's just that if we were sub-classing `AssetLocator` for the usecase of supporting multiple different asset types in `run_heuristics` then I'm afraid we need to change it because it's not recommended in the documentation. But as you say that we are not currently doing it that way, then I don't see any major issue with sub-classing `AssetLocator` to ensure `PyPIPackageJsonAsset` is in the right "shape".

Reply:
I noticed that the current implementation is recreating a new `PyPIRegistry` instead of using the detected one from `ctx.dynamic_data["package_registries"]`. This change fixes this issue and also stores the metadata in the `PackageRegistryInfo` object, which also enforces type checking. So, no need for sub-typing anymore.

See commit [a2b0821](https://github.com/oracle/macaron/pull/797/commits/a2b08218e5ac2e67c6d5d5095e3707690f743bce).